### PR TITLE
doc/radosgw/d3n: add user docs

### PR DESCRIPTION
**Pull request by mkogan1**
(high-level feature description, setup instructions, config options)

Fixes: https://tracker.ceph.com/issues/51409

Signed-off-by: Mark Kogan <mkogan@redhat.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/radosgw/d3n_datacache.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+| The Rados Gateways act as cache servers for the back-end object store (OSDs), storing data locally for reuse.
+
+
+See `MOC D3N (Datacenter-scale Data Delivery Network)`_ and `Red Hat Research D3N Cache for Data Centers`_.
+
+Implementation
+==============
+
+- D3N cache supports both the `S3` and `Swift` object storage interfaces.
+- D3N will cache tail objects (by default it is parts of objects that are larger than 4MB).
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:29]: because they are immutable

File: doc/radosgw/d3n_datacache.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+Requirements
+------------
+
+- A SSD(/dev/nvme,/dev/pmem,/dev/shm) storage device XFS formatted and mounted, it will be used as the cache backing store.
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:34]: how does rgw know the device is xfs-formatted?  does that matter?

File: doc/radosgw/d3n_datacache.rst [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  (depending on device performance, multiple RGWs may share a single device but each requires a discrete directory on said device)
+- The `rgw_enable_ops_log` config variable needs be enabled for D3N operation.
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:36]: why?  what is the connection with ops-log?

File: doc/radosgw/d3n_datacache.rst [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+Limitations
+-----------
+
+- D3N will not cache objects compressed by `Rasdos Gtateway Compression`_ (OSD level compression is supported).
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:41]: gtateway

**Review by mattbenjamin — COMMENTED** [review]
looks pretty good, couple of questions;  maybe there should be more on architecture?

File: doc/radosgw/d3n_datacache.rst (hunk 1 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: added

File: doc/radosgw/d3n_datacache.rst (hunk 2 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: not required, just only tested with XFS, changing wording accordingly

File: doc/radosgw/d3n_datacache.rst (hunk 4 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: fixed

File: doc/radosgw/d3n_datacache.rst (hunk 3 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst:36]: was a requirement of the original PR, believe no longer necessary, have a task to re-check, will update

Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst:36]: checked & not necesary, will remove

File: doc/radosgw/d3n_datacache.rst [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+than the bandwidth of a solid-state storage located at an edge node.
+
+| D3N improves the performance of big-data jobs running in analysis clusters by speeding up recurring reads from the data lake.
+| The Rados Gateways act as cache servers for the back-end object store (OSDs), storing data locally for reuse.
+
+Architecture
+============
+
+D3N improves the performance of big-data jobs by speeding up repeatedly accessed dataset reads from the data lake.
+Cache servers are located in the datacenter on the access side of potential network and storage bottlenecks,
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:26]: full stop/period after bottlenecks

File: doc/radosgw/d3n_datacache.rst [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+D3Ns two layer logical caches forms a traditional caching hierarchy :sup:`*`
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:27]: two-layer

Comment (mattbenjamin) [reply doc/radosgw/d3n_datacache.rst:27]: caches form ("caches" is plural)

File: doc/radosgw/d3n_datacache.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+by a request to the data lake (Rados)
+
+:sup:`*` currently only layer 1 cache has been upstreamed.
+
+See `MOC D3N (Datacenter-scale Data Delivery Network)`_ and `Red Hat Research D3N Cache for Data Centers`_.
+
+Implementation
+==============
+
+- D3N cache supports both the `S3` and `Swift` object storage interfaces.
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:43]: drop "cache" here, or else write "The D3N cache"

File: doc/radosgw/d3n_datacache.rst [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+- D3N will cache tail objects because they are immutable (by default it is parts of objects that are larger than 4MB).
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:44]: D3N currently caches only tail objects, because

File: doc/radosgw/d3n_datacache.rst [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+Requirements
+------------
+
+- A SSD(/dev/nvme,/dev/pmem,/dev/shm) storage device, formatted (filesystems other than XFS were not tested) and mounted,
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:49]: An SSD (...) or similar block storage device, filesystem (...

File: doc/radosgw/d3n_datacache.rst [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  it will be used as the cache backing store.
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:50]: Full stop/period after "mounted."  Capitalize "It"

File: doc/radosgw/d3n_datacache.rst [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    -o rgw_d3n_l1_datacache_size=10737418240
+
+the above example assumes that the cache backing-store solid state device
+is mounted at `/mnt/nvme0` and has `10 GB` of free space available.
+
+    NOTE: Each time the Rados Gateway is restarted the content of the cache directory is purged.
+
+Logs
+----
+- D3N related log lines in `radosgw.*.log` contain the string `d3n` (case insensitive).
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:81]: in what way is "d3n" case insensitive in a log line?

Comment (mattbenjamin) [reply doc/radosgw/d3n_datacache.rst:81]: i.e, case insensitivity can be a property of a filesystem or collation system, but not of a string literal

File: doc/radosgw/d3n_datacache.rst [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+| ``rgw_d3n_l1_datacache_persistent_path``           | string   | path of the directory for storing the local cache objects data                                 |
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+| ``rgw_d3n_l1_datacache_size``                      | size     | datacache maximum size on disk in bytes                                                        |
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+| ``rgw_d3n_l1_eviction_policy``                     | enum     | select the d3n cache eviction policy, allowed values: (`lru`|`random`)                         |
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+
+
```
Comment (mattbenjamin) [inline doc/radosgw/d3n_datacache.rst:102]: should there be a "References" section heading?

File: doc/radosgw/d3n_datacache.rst [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+| ``rgw_d3n_l1_local_datacache_enabled``             | bool     | Enable datacenter-scale dataset delivery local cache                                           |
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+| ``rgw_d3n_l1_datacache_persistent_path``           | string   | path of the directory for storing the local cache objects data                                 |
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+| ``rgw_d3n_l1_datacache_size``                      | size     | datacache maximum size on disk in bytes                                                        |
++----------------------------------------------------+----------+------------------------------------------------------------------------------------------------+
+| ``rgw_d3n_l1_eviction_policy``                     | enum     | select the d3n cache eviction policy, allowed values: (`lru`|`random`)                         |
```
Comment (cbodley) [inline doc/radosgw/d3n_datacache.rst:99]: i think we should prefer the yaml integration for these, so these docs are automatically updated to match the source:
```rst
.. confval:: rgw_d3n_l1_local_datacache_enabled
.. confval:: rgw_d3n_l1_datacache_persistent_path
.. confval:: rgw_d3n_l1_datacache_size
.. confval:: rgw_d3n_l1_eviction_policy
```
for an example of how these render, see https://docs.ceph.com/en/latest/radosgw/config-ref/

**Review by cbodley — COMMENTED** [review]
👍

File: doc/radosgw/d3n_datacache.rst (hunk 11 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst:81]: some of the logs have `... D3N ...` and some of the logs `... d3n ...`, will add to the D3N talk list a log cleanup PR

File: doc/radosgw/d3n_datacache.rst (hunk 12 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst:108]: added to "CONFIG REFERENCE" section like in other docs

File: doc/radosgw/d3n_datacache.rst (hunk 13 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: very cool, done, thanks

File: doc/radosgw/d3n_datacache.rst [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+Running
+-------
+
+D3N minimal configuration startup example using `vstart.sh`::
+
+    env MON=1 OSD=1 MDS=0 MGR=1 RGW=1 ../src/vstart.sh -n  \
+    -o rgw_d3n_l1_local_datacache_enabled=true  \
+    -o rgw_d3n_l1_datacache_persistent_path=/mnt/nvme0/rgw_datacache/  \
+    -o rgw_d3n_l1_datacache_size=10737418240
```
[the remark this answers is not in the document]
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: changed the instructions/examples to instruct how to enable D3N by modifying ceph.conf
and added a note/disclaimer about possible cache data duplication in multiple RGWs co-location configuration.

**Comment by mkogan1** [comment]
@mattbenjamin @cbodley please check the changes addressing the review comments in the 2nd commit
(did not squash yet intentionally)

File: doc/radosgw/d3n_datacache.rst (hunk 5 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: ack

File: doc/radosgw/d3n_datacache.rst (hunk 6 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: ack

File: doc/radosgw/d3n_datacache.rst (hunk 7 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: ack

File: doc/radosgw/d3n_datacache.rst (hunk 8 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: ack

File: doc/radosgw/d3n_datacache.rst (hunk 9 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: ack

File: doc/radosgw/d3n_datacache.rst (hunk 10 above)
Comment (mkogan1) [reply doc/radosgw/d3n_datacache.rst]: ack

**Comment by cbodley** [comment]
jenkins render docs

**Comment by cbodley** [comment]
hmm we also have a "Data caching and CDN" page at https://ceph--42340.org.readthedocs.build/en/42340/radosgw/rgw-cache/

i imagine it'll be confusing to see two completely separate pages for data caching in the docs, though i'm not sure exactly how to frame their comparison

**Comment by mkogan1** [comment]
> hmm we also have a "Data caching and CDN" page at https://ceph--42340.org.readthedocs.build/en/42340/radosgw/rgw-cache/
> 
> i imagine it'll be confusing to see two completely separate pages for data caching in the docs, though i'm not sure exactly how to frame their comparison

yes, it's the previous iteration on caching that leverages the NGINX internal cache mechanism to provide an external data cache.
currently, the biggest difference is that the NGINX cache supports caching of head objects and the D3N does not as of yet.
adding a note to both.

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
lgtm. care to re-review @mattbenjamin?

**Review by mattbenjamin — APPROVED** [review]
lgtm, squash away

**Pull request merged**